src/broadcast.jl

```diff
@@ -1,158 +1,110 @@
-    A
-end
-@inline function broadcast_t(f, T::Type{Bool}, shape, it, A::GPUArrays.GPUArray, Bs::Vararg{Any,N}) where N
-    C = similar(A, T, shape)
-    keeps, Idefaults = map_newindexer(shape, A, Bs)
-    _broadcast!(f, C, keeps, Idefaults, A, Bs, Val{N}, it)
-    return C
-end
-@inline function broadcast_t(f, T::Type{Bool}, shape, it, A::GPUArrays.GPUArray, B::GPUArrays.GPUArray, Bs::Vararg{Any,N}) where N
-    C = similar(A, T, shape)
-    Bs = (B, Bs...)
-    keeps, Idefaults = map_newindexer(shape, A, Bs)
-    _broadcast!(f, C, keeps, Idefaults, A, Bs, Val{N}, it)
-    return C
-end
+import Base.Broadcast: BroadcastStyle, AbstractArrayStyle, Broadcasted, broadcast_axes
+import Base.Broadcast: DefaultArrayStyle, materialize!, flatten, ArrayStyle, combine_styles
 
-@inline function broadcast_t(
-        f, ::Type{T}, shape, iter, A::GPUArray, Bs::Vararg{Any,N}
-    ) where {N, T}
-    C = similar(A, T, shape)
-    keeps, Idefaults = map_newindexer(shape, A, Bs)
-    _broadcast!(f, C, keeps, Idefaults, A, Bs, Val{N}, iter)
-    return C
-end
-@inline function broadcast_t(
-        f, ::Type{T}, shape, iter, A::GPUArray, B::GPUArray, rest::Vararg{Any,N}
-    ) where {N, T}
-    C = similar(A, T, shape)
-    Bs = (B, rest...)
-    keeps, Idefaults = map_newindexer(shape, A, Bs)
-    _broadcast!(f, C, keeps, Idefaults, A, Bs, Val{N}, iter)
-    return C
+BroadcastStyle(::Type{T}) where T <: GPUArray = ArrayStyle{T}()
+BroadcastStyle(::Type{Any}, ::Type{T}) where T <: GPUArray = ArrayStyle{T}()
+BroadcastStyle(::Type{T}, ::Type{Any}) where T <: GPUArray = ArrayStyle{T}()
+BroadcastStyle(::Type{T1}, ::Type{T2}) where {T1 <: GPUArray, T2 <: GPUArray} = ArrayStyle{T}()
+
+const GPUBroadcast = Broadcasted{<: ArrayStyle{<: GPUArray}}
+
+function Base.similar(bc::Broadcasted{ArrayStyle{GPU}}, ::Type{ElType}) where {GPU <: GPUArray, ElType}
+    similar(GPU, ElType, axes(bc))
 end
 
-@inline function broadcast_t(
-        f, T, shape, iter, A::Any, B::GPUArray, rest::Vararg{Any, N}
-    ) where N
-    C = similar(B, T, shape)
-    Bs = (B, rest...)
-    keeps, Idefaults = map_newindexer(shape, A, Bs)
-    _broadcast!(f, C, keeps, Idefaults, A, Bs, Val{N}, iter)
-    return C
+# copy overload
+
+function gpu_eachindex(f, axes, A::GPUArray, args...)
+    # TODO use axes + offset etc
+    shape = length.(axes)
+    gpu_call(A, (f, shape, A, args...), prod(shape)) do state, f, A, args...
+        lidx = linear_index(state)
+        lidx > length(A) && return
+        cartesian = CartesianIndex(gpu_ind2sub(shape, lidx))
+        f(state, cartesian, A, args...)
+        return
+    end
 end
-function broadcast_t(f::Any, ::Type{Any}, ::Any, ::Any, A::GPUArrays.GPUArray, args::Vararg{Any,N}) where N
-    error("Return type couldn't be inferred for broadcast. Func: $f, $(typeof(A)), $args")
+
+# copyto! overloads
+@inline function copyto!(dest::GPUArray, B::GPUBroadcast)
+    flat = flatten(B); as = flat.args; f = flat.f
```

Comment:
You don't need to flatten the Broadcast as long as you can convert both `Broadcasted` and `Extruded` to GPU representations:

```
function CUDAnative.cudaconvert(bc::Broadcast.Broadcasted{Style}) where Style
    Broadcast.Broadcasted{Style}(bc.f, map(CUDAnative.cudaconvert, bc.args), bc.axes)
end

function CUDAnative.cudaconvert(ex::Broadcast.Extruded)
    Broadcast.Extruded(CUDAnative.cudaconvert(ex.x), ex.keeps, ex.defaults)
end
```

which simplifies broadcast to:

```
function _cuda_copyto_kernel!(dest::CuDeviceArray, bc::Broadcast.Broadcasted{Nothing})
    shape = axes(dest)
    let I = @cuda_index(shape)
        @inbounds dest[I] = bc[I]
    end
end
```